cocotb/share/lib/gpi_log/gpi_logging.c

```diff
@@ -146,53 +156,65 @@ void gpi_log(const char *name, long level, const char *pathname, const char *fun
 
     // Ignore truncation
-    // calling args is level, filename, lineno, msg, function
-    //
     PyGILState_STATE gstate = PyGILState_Ensure();
 
-    PyObject *check_args = PyTuple_New(1);
-    PyTuple_SetItem(check_args, 0, PyLong_FromLong(level));
+    PyObject *pLevel = PyLong_FromLong(level);                  // New reference
+    if (pLevel == NULL)
+        goto error;
 
-    PyObject *filter_ret = PyObject_CallObject(pLogFilter, check_args);
-    Py_DECREF(check_args);
-    if (filter_ret == NULL) {
-        PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
-    }
-    int is_enabled = PyObject_IsTrue(filter_ret);
-    Py_DECREF(filter_ret);
+    PyObject *pFilterReturn = PyObject_CallFunctionObjArgs(pLogFilter, pLevel, NULL);
+    if (pFilterReturn == NULL)
+        goto error;
+
+    int is_enabled = PyObject_IsTrue(pFilterReturn);
+    Py_DECREF(pFilterReturn);
     if (is_enabled < 0) {
-        /* A python exception occured while converting `filter_ret` to bool */
-        PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
+        /* A python exception occured while converting `pFilterReturn` to bool */
+        goto error;
     }
 
     if (!is_enabled) {
-        PyGILState_Release(gstate);
-        return;
+        goto ok;
     }
 
     va_start(ap, msg);
     n = vsnprintf(log_buff, LOG_SIZE, msg, ap);
     va_end(ap);
 
-    PyObject *call_args = PyTuple_New(5);
-    PyTuple_SetItem(call_args, 0, PyLong_FromLong(level));           // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 1, PyUnicode_FromString(pathname));   // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 2, PyLong_FromLong(lineno));          // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 3, PyUnicode_FromString(log_buff));   // Note: This function steals a reference.
-    PyTuple_SetItem(call_args, 4, PyUnicode_FromString(funcname));
-
-    PyObject *handler_ret = PyObject_CallObject(pLogHandler, call_args);
-    Py_DECREF(call_args);
-    if (handler_ret == NULL){
-        PyErr_Print();
-        PyGILState_Release(gstate);
-        return;
-    }
-    Py_DECREF(handler_ret);
-
+    PyObject *pFileName = PyUnicode_FromString(pathname);       // New reference
+    if (pFileName == NULL)
+        goto error;
+
+    PyObject *pLineNo = PyLong_FromLong(lineno);                // New reference
+    if (pLineNo == NULL)
+        goto error;
+
+    PyObject *pMessage = PyUnicode_FromString(log_buff);        // New reference
+    if (pMessage == NULL)
+        goto error;
+
+    PyObject *pFunctionName = PyUnicode_FromString(funcname);   // New reference
+    if (pFunctionName == NULL)
+        goto error;
+
+    Py_INCREF(pLevel);
+
+    // Log function args are level, filename, lineno, msg, function
+    PyObject *pLogReturn = PyObject_CallFunctionObjArgs(pLogHandler, pLevel, pFileName, pLineNo, pMessage, pFunctionName, NULL);
+    Py_DECREF(pLevel);
```

Comment:
This line seems a little pointless, since you decref all the variables down below anyway.

Reply:
You're correct. I was still thinking of how I did it before, where the tuple argument to pLogFilter stole the reference. Removed the extra INCREF/DECREF pair.